SharedMemory and NDArray implementation

tpietzsch opened the pull request:
This PR adds 

**1)**  `SharedMemory` which is adapted from @carlosuc3m code in [JDLL](https://github.com/bioimage-io/JDLL/).
The Linux implementation has been tested only superficially, the Windows implementation not at all ...

**2)** `NDArray` which annotates `SharedMemory` with data type and shape, similar to NumPy `ndarray`.
This is independent of ImgLib2, adapters for ImgLib2 data structures are in https://github.com/imglib/imglib2-appose.

**3)** JSON de/serialization for `SharedMemory` and `NDArray`. Basically: put a `NDArray` into a `Task`'s`inputs` on the Java side and a `NDArray` comes out in the `GroovyWorker` inputs, or a NumPy `ndarray` in the python worker inputs.
(The python side requires the corresponding https://github.com/apposed/appose-python/pull/1 PR)

tpietzsch committed:
Add SharedMemory and NDArray

SharedMemory implementation is lifted from JDLL.
NDArray wraps SharedMemory, and adds shape and data-type information.

Co-authored-by: Carlos Garcia Lopez de Haro <100329787@alumnos.uc3m.es>

tpietzsch committed:
Add Python example with NDArray

(requires branch 'schmarrn' in appose-python to work)

tpietzsch committed:
Add C src for ShmCreate.dylib

Co-authored-by: Carlos Garcia Lopez de Haro <100329787@alumnos.uc3m.es>

ctrueden reviewed (commented):
Thank you so much @tpietzsch for bringing this to completion. It's great. I am only going to make some small adjustments, which I will list here as part of my review, in case you have any concerns about any of them:

1. Reformat Java code to use tabs in line with the existing code.
2. Add class Javadoc to all classes, including proper `@author` attribution, as well as references to analogous Python API for comparison.
3. Rewrite the main commit adding the implementation to have a `Co-authored-by` line crediting @carlosuc3m.
4. Rename `ShmImpl` to just `Shm`, because the `Impl` suffix in Java classes is traditionally used to indicate an implementation of an interface, not the interface itself.
5. To avoid the `org.apposed.appose` base package referencing the `org.apposed.appose.shm` subpackage: move the `ShmImpl`/`Shm` interface out of the `Shm` subpackage, and alter its `create` static method to use `ServiceLoader` instead with platform-specific `Shm` object factories. Details TBD.

I'll also look into adding a `SharedMemoryTest` to the unit tests, and broadening the GitHub Actions CI to run on all supported platforms, since this is critical foundational logic.

ctrueden committed:
Refactor platform-specific shm logic to plugins

This makes the SharedMemory class into an interface with three
implementations, one per supported platform, but leaves the door open
for additional platforms to add support by implementing their own
ShmFactory that produces appropriate SharedMemory objects.

The plugins are discovered using Java's ServiceLoader mechanism.

ctrueden committed:
Add simpler SharedMemory creation methods

There are only two cases: create or attach.

ctrueden committed:
Add unit tests for SharedMemory functionality

ctrueden committed:
POM: credit people who contributed

ctrueden committed:
Make DType enum a static inner class of NDArray

The DType and Shape classes define the two major pieces of metadata
needed by NDArray. One was an inner class of NDArray and the other
was not. This change makes it consistent: now both are inner classes.

ctrueden committed:
Add tests for the encode/decode methods of Types

ctrueden committed:
Migrate common shm size checking logic to helper

ctrueden committed:
Test NDArray<->JSON en/de-coding

ctrueden committed:
Make shm the third, not first, NDArray attribute

So that it matches the Python implementation of Appose.
The reason it's third in Python is so that it can have the default
parameter value of None, to provide feature parity with the Java
version's two-argument NDAray constructor that creates the SharedMemory.

And create the memory even if the three-argument constructor is used.

ctrueden committed:
Do some minor cleanups

ctrueden committed:
Remove blanket `throws Exception` declarations

ctrueden committed:
Update license headers for 2024

Thanks, license-maven-plugin!

ctrueden committed:
Split common Shm logic to base class

And decouple the close and unlink functions.

There was a lot of duplicate code between ShmLinux, ShmMacOS,
and ShmWindows, which is now in the internal ShmBase class.

This made it easier to have consistent logic around the close
vs. unlink behavior across platforms, with less copy/pasted code.
